apps/libsignpost/port_signpost_arduino.cpp

```diff
@@ -0,0 +1,252 @@
+	Wire.onRequest(slave_read_callback);
+	return 0;
+}
+
+int port_signpost_i2c_master_write(uint8_t addr, uint8_t* buf, size_t len) {
+	if (len == 0) {
+		return -1;
+	}
+	//Initialize i2c as master
+	Wire.begin();
+	//Set i2c clock speed
+	Wire.setClock(SIGNPOST_I2C_SPEED);
+	//Send i2c message to addr
+	Wire.beginTransmission(addr);
+	int num_written = Wire.write(buf, len);
+	int rc = Wire.endTransmission();
+	//Put Arduino back into i2c slave mode
+	Wire.begin(g_arduino_i2c_address);
+	return num_written;
+}
+
+int port_signpost_i2c_slave_listen(port_signpost_callback cb, uint8_t* buf, size_t max_len) {
+	//Initialize i2c as slave
+	Wire.begin(g_arduino_i2c_address);
+	//Set i2c slave listen callback and buffer
+	g_slave_listen_callback = cb;
+	g_slave_receive_buf = buf;
+	g_slave_receive_buf_max_len = max_len;
+	g_slave_receive_buf_len = 0;
+	return 0;
+}
+
+//Function untested
+int port_signpost_i2c_slave_read_setup(uint8_t* buf, size_t len) {
+	//Initialize i2c as slave
+	Wire.begin(g_arduino_i2c_address);
+	//Set global i2c slave transmit buffer
+	g_slave_transmit_buf = buf;
+	g_slave_transmit_buf_len = len;
+	return 0;
+}
+
+int port_signpost_mod_out_set() {
+	digitalWrite(ARDUINO_MOD_OUT, HIGH);
+	return 0;
+}
+
+int port_signpost_mod_out_clear() {
+	digitalWrite(ARDUINO_MOD_OUT, LOW);
+	return 0;
+}
+
+int port_signpost_mod_in_read() {
+	return digitalRead(ARDUINO_MOD_IN);
+}
+
+int port_signpost_mod_in_enable_interrupt_falling(port_signpost_callback cb) {
+	//Set global MOD_IN falling edge callback 
+	g_mod_in_callback_falling = cb;
+	//Attach callback helper to hardware interrupt
+	attachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN), mod_in_callback_falling_helper, FALLING);
+	return 0;
+}
+
+int port_signpost_mod_in_enable_interrupt_rising(port_signpost_callback cb) {
+	//Set global MOD_IN rising edge callback 
+	g_mod_in_callback_rising = cb;
+	//Attach callback helper to hardware interrupt
+	attachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN), mod_in_callback_rising_helper, RISING);
+	return 0;
+}
+
+int port_signpost_mod_in_disable_interrupt(void) {
+	detachInterrupt(digitalPinToInterrupt(ARDUINO_MOD_IN));
+	return 0;
+}
+
+//TODO: Implement version that puts Arduino into a sleep state
+void port_signpost_wait_for(void* wait_on_true) {
+	volatile bool wait_on_true_volatile = false;
+	//Stay in loop until wait_on_true becomes true
+	while (!wait_on_true_volatile) {
+		wait_on_true_volatile = *((bool*) wait_on_true);
+	}
+}
+
+//TODO: Implement version that actually times out
+//		For now it is just a replica of port_signpost_wait_for
+int port_signpost_wait_for_with_timeout(void* wait_on_true, uint32_t ms) {
+	port_signpost_wait_for(wait_on_true);
+	return SB_PORT_SUCCESS;
+}
+
+void port_signpost_delay_ms(unsigned ms) {
+	//For every 1 ms delay, call delayMicroseconds once.
+	//delayMicroseconds does not work for inputs > 16383,
+	//so a loop is used to divide up the duration.
+	for (unsigned i = 0; i < ms; ++i) {
```

Comment:
We should eventually do something a bit smarter here. We could partition based on the largest counter value instead of 1000. This would provide more accurate delays. This change isn't needed for the merge though.